app/cdap/components/PipelineHistory/PipelineHistoryStore.tsx

```diff
@@ -0,0 +1,197 @@
+
+interface IState {
+  previousTokens: string[];
+  nextPageToken: string;
+  pageToken: string;
+  pageLimit: number;
+  pipelineVersions: IPipelineVersion[];
+  ready: boolean;
+  pageLimitOptions: number[];
+}
+
+interface IStore {
+  versions: IState;
+}
+
+export const QUERY = gql`
+  query Query(
+    $namespace: String
+    $pageSize: Int
+    $token: String
+    $nameFilter: String
+    $orderBy: String
+    $nameFilterType: String
+  ) {
+    pipelines(
+      namespace: $namespace
+      pageSize: $pageSize
+      pageToken: $token
+      nameFilter: $nameFilter
+      orderBy: $orderBy
+      nameFilterType: $nameFilterType
+    ) {
+      applications {
+        name
+        version
+        artifact {
+          name
+        }
+        runs {
+          status
+          starting
+        }
+        totalRuns
+        nextRuntime {
+          id
+          time
+        }
+        change {
+          description
+          creationTimeMillis
+        }
+      }
+      nextPageToken
+    }
+  }
+`;
+
+const Actions = {
+  prevPage: 'PIPELINE_VERSIONS_PREV_PAGE',
+  nextPage: 'PIPELINE_VERSIONS_NEXT_PAGE',
+  reset: 'PIPELINE_VERSIONS_RESET',
+  setVersions: 'PIPELINE_VERSIONS_SET_VERSIONS',
+  updateVersions: 'PIPELINE_VERSIONS_UPDATE_PIPELINES',
+  setPageLimit: 'PIPELINE_VERSIONS_SET_PAGE_LIMIT',
+};
+
+const defaultInitialState: IState = {
+  nextPageToken: null,
+  previousTokens: [],
+  pageToken: null,
+  pageLimit: 5,
+  pipelineVersions: [],
+  ready: false,
+  pageLimitOptions: [4, 5, 6, 7, 8, 9],
+};
+
+const versions: Reducer<IState> = (state = defaultInitialState, action: IAction) => {
+  switch (action.type) {
+    case Actions.prevPage:
+      const lastTokenIdx = state.previousTokens.length - 1;
+      const lastToken = state.previousTokens[lastTokenIdx];
+      return {
+        ...state,
+        previousTokens: state.previousTokens.slice(0, lastTokenIdx),
+        nextPageToken: state.pageToken,
+        pageToken: lastToken,
+        hasMultiple: true,
+        ready: false,
+      };
+    case Actions.nextPage:
+      return {
+        ...state,
+        previousTokens: [...state.previousTokens, state.pageToken],
+        nextPageToken: null,
+        pageToken: state.nextPageToken,
+        hasMultiple: true,
+        ready: false,
+      };
+    case Actions.setVersions:
+      return {
+        ...state,
+        pipelineVersions: action.payload.pipelineVersions,
+        nextPageToken: action.payload.nextPageToken,
+        ready: true,
+      };
+    case Actions.updateVersions:
+      return {
+        ...state,
+        pipelineVersions: action.payload.pipelineVersions,
+      };
+    case Actions.reset:
+      return defaultInitialState;
+    case Actions.setPageLimit:
+      return {
+        ...defaultInitialState,
+        pageLimit: action.payload.pageLimit,
+      };
+    default:
+      return state;
+  }
+};
+
+const Store: StoreInterface<IStore> = createStore(
+  combineReducers({
```

Comment:
was the type complaining here or sometrhint

Reply:
no complain